### PR TITLE
Check guest health before and after each virt feature test

### PR DESCRIPTION
**Pull request by Julie-CAO**
To fix the following problems:

- It is difficult to determine which tests is the culprit of errors or warnings in journals.

- In addition, current uploaded log by "journalctl -b" does not include the log of previous boot which reports the errors. fg. https://openqa.suse.de/tests/13564742#downloads => libvirt_virtual_network_init-journalctl-sles-15-sp6-64-fv-uefi-xen.log

- Related ticket: https://progress.opensuse.org/issues/155728
- Verification run: 
[gi-guest_sles15sp5-on-host_developing-xen](https://openqa.suse.de/tests/13763065)
[gi-host_sles12sp5-xen](http://10.67.129.27/tests/291)
[uefi-gi-guest_developing-on-host_sles12sp5-xen](https://openqa.suse.de/tests/13763082)
[gi-guest_sles12sp5-on-host_developing-xen](https://openqa.suse.de/tests/13763066)
[a failed uefi-gi-guest_developing-on-host_developing-xen](http://openqa.suse.de/tests/13722945)
[a failed sriov test on uefi guest](http://10.67.129.27/tests/318)
[continue feature tests on on-host_developing-xen above](http://10.67.129.27/tests/326)

### CODE REVIEW

**Comment by Julie-CAO** [comment]
@alice-suse @waynechen55 @guoxuguang and others, welcome review!

File: lib/virt_feature_test_base.pm
Comment (alice-suse) [inline lib/virt_feature_test_base.pm:76]: Will running  check_guest_health *once* here for each feature test be enough to tell where the journal problem happens?
BTW, may I ask how long in general will it take for adding this check?

Comment (Julie-CAO) [reply lib/virt_feature_test_base.pm:76]: > Will running check_guest_health once here for each feature test be enough to tell where the journal problem happens?

No, it is hard to determine which test owns the error because the guest is rebooted in each test. In addition, only the jouranls of current boot will be uploaded, we missed the log as in the PR description.

Comment (Julie-CAO) [reply lib/virt_feature_test_base.pm:76]: > BTW, may I ask how long in general will it take for adding this check?

It takes 42 seconds in a case.

```
[2024-02-22T16:39:56.946771Z] [debug] [pid:32331] tests/virt_autotest/unified_guest_installation.pm:103 called virt_autotest::utils::check_guest_health -> lib/virt_autotest/utils.pm:319 called testapi::script_run
...
[2024-02-22T16:40:38.913179Z] [debug] [pid:32331] tests/virt_autotest/unified_guest_installation.pm:103 called virt_autotest::utils::check_guest_health -> lib/virt_autotest/utils.pm:331 called testapi::record_info
[2024-02-22T16:40:38.913389Z] [debug] [pid:32331] <<< testapi::record_info(title="Healthy guest!", output="sles-15-sp6-64-fv-uefi-xen looks good so far!", result="ok")
```

Comment (alice-suse) [reply lib/virt_feature_test_base.pm:76]: Thank you for the data.
Is it correct understanding that this PR is not *complete* fix for the problems you described yet?

Comment (alice-suse) [reply lib/virt_feature_test_base.pm:76]: Thank you for the data.
Is it correct understanding that this PR is not *complete* fix for the problems you described yet?

File: lib/virt_autotest/utils.pm
Comment (waynechen55) [inline lib/virt_autotest/utils.pm:326]: I think this block can be simplified using just ```elsif``` instead of ```if``` in a ```else```.

File: lib/virt_feature_test_base.pm
Comment (Julie-CAO) [reply lib/virt_feature_test_base.pm:76]: It does fix the two problems in the PR description.

File: lib/virt_autotest/utils.pm
Comment (Julie-CAO) [reply lib/virt_autotest/utils.pm:326]: Yes, Thanks.

**Comment by guoxuguang** [comment]
LGTM

**Review by alice-suse — APPROVED** [review]
LGTM, please fix the minor check issues

**Comment by Julie-CAO** [comment]
The final version is ready which include more commits. The verification runs were updated in PR description. Welcome review!  @alice-suse @waynechen55 @guoxuguang @RoyCai7

File: lib/virt_autotest/utils.pm
Comment (alice-suse) [inline lib/virt_autotest/utils.pm:258]: @Julie-CAO  Hi Julie, thanks for the enhancement. The logic of function check_failures_in_journal and check_failures_in_journal seems very complex now. Would you please add some comment explaining the overall design/idea of the two functions? It will help maintainers understand code easier in future.

Comment (Julie-CAO) [reply lib/virt_autotest/utils.pm:258]: > The logic of function check_failures_in_journal and check_failures_in_journal seems very complex now. Would you please add some comment explaining the overall design/idea of the two functions?

Hi @alice-suse , which two functions? I can add some description at the beginning of `sub check_failures_in_journal()`

Comment (alice-suse) [reply lib/virt_autotest/utils.pm:267]: oh, sorry, one function only. I noticed you added some explanation. It is much clearer now. Thanks.

File: lib/virt_autotest/utils.pm
Comment (alice-suse) [inline lib/virt_autotest/utils.pm:258]: Typo? imcremenal -> incremental?

Comment (Julie-CAO) [reply lib/virt_autotest/utils.pm]: Corrected. thanks.

**Comment by alice-suse** [comment]
LGTM

**Comment by Julie-CAO** [comment]
Is this ok to be merged? @alice-suse

**Comment by alice-suse** [comment]
I was waiting for other's review. Seems no more comment. Let's merge it then. Thanks for the great PR!

**Pull request merged**
